core/src/main/java/io/questdb/griffin/engine/functions/groupby/FirstStrGroupByFunction.java

```diff
@@ -32,33 +32,70 @@
 import io.questdb.griffin.engine.functions.GroupByFunction;
 import io.questdb.griffin.engine.functions.StrFunction;
 import io.questdb.griffin.engine.functions.UnaryFunction;
-import io.questdb.std.str.StringSink;
+import io.questdb.std.Misc;
+import io.questdb.std.ObjList;
+import io.questdb.std.str.DirectCharSink;
 import org.jetbrains.annotations.NotNull;
 
 public class FirstStrGroupByFunction extends StrFunction implements GroupByFunction, UnaryFunction {
+
+    private static final int INITIAL_SINK_CAPACITY = 16;
+    private static final int LIST_CLEAR_THRESHOLD = 64;
+
     protected final Function arg;
-    private final StringSink sink = new StringSink();
+    protected final ObjList<DirectCharSink> sinks = new ObjList<>();
+
     protected int valueIndex;
+    private int sinkIndex = 0;
 
     public FirstStrGroupByFunction(@NotNull Function arg) {
         this.arg = arg;
     }
 
     @Override
     public void clear() {
-        sink.clear();
+        // Free extra sinks.
+        if (sinks.size() > LIST_CLEAR_THRESHOLD) {
```

Comment:
nit: I think this would be slightly better

```Java
        for (int i = LIST_CLEAR_THRESHOLD, n = sinks.size(); i < n; i++) {
            sinks.getAndSetQuick(i, null).close();
        }
        sinks.setPos(LIST_CLEAR_THRESHOLD);
```

Reply:
applied in latest commit